### PR TITLE
Service Registry ping and ttl

### PR DESCRIPTION
**Pull request by borditamas**

### CODE REVIEW

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	
+	@Autowired
+	private ServiceRegistryDBService serviceRegistryDBService;
+	
+	@Value (CommonConstants.$SERVICE_REGISTRY_PING_TIMEOUT_WD)
+	private int timeout;
+	
+	@Override
+	public void execute(final JobExecutionContext context) throws JobExecutionException {
+		logger.info("STARTED: Providers reachability task");
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java]: use debug

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		List<Long> removedServiceRegistryIDs = checkProvidersReachability();
+		logger.info("FINISHED: Providers reachability task. Number of removed service registry entry: " + removedServiceRegistryIDs.size());
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java]: and here too

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+				final int totalPages = pageOfServiceEntries.getTotalPages();
+				removedServiceRegistryIDs.addAll(pingAndRemoveRegisteredServices(pageOfServiceEntries));
+				pageIndexCounter++;
+				while (pageIndexCounter < totalPages) {
+					pageOfServiceEntries = serviceRegistryDBService.getAllServiceReqistryEntries(pageIndexCounter, pageSize, Direction.ASC, CommonConstants.COMMON_FIELD_NAME_ID);
+					removedServiceRegistryIDs.addAll(pingAndRemoveRegisteredServices(pageOfServiceEntries));
+					pageIndexCounter++;
+				}
+			}
+		} catch (IllegalArgumentException exception) {
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java]: final

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	
+	private List<Long> pingAndRemoveRegisteredServices(final Page<ServiceRegistry> pageOfServiceEntries) {
+		final List<Long> removedServiceRegistryIDs = new ArrayList<>();
+		for (final ServiceRegistry serviceRegistryEntry : pageOfServiceEntries) {
+			final System provider = serviceRegistryEntry.getSystem();
+			final String address = provider.getAddress();
+			final int port = provider.getPort();
+			if (! pingService(address, port)) {
+				serviceRegistryDBService.removeServiceRegistryEntryById(serviceRegistryEntry.getId());
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTask.java]: consider deleting multiple entries at once

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTaskConfig.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			logger.info("Providers reachabilitiy task is not adjusted");
+		}
+		return schedulerFactory;        
+    }
+	
+	@Bean
+    public SimpleTriggerFactoryBean providersReachabilityTaskTrigger() {
+		final SimpleTriggerFactoryBean trigger = new SimpleTriggerFactoryBean();
+		trigger.setJobDetail(providersReachabilityTaskDetail().getObject());
+        trigger.setRepeatInterval(pingInterval * 60000);
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTaskConfig.java]: constant maybe

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/AutoWiringSpringBeanQuartzTaskFactory.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package eu.arrowhead.core.serviceregistry.quartz;
+
+import org.quartz.spi.TriggerFiredBundle;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.scheduling.quartz.SpringBeanJobFactory;
+
+public final class AutoWiringSpringBeanQuartzTaskFactory extends SpringBeanJobFactory implements ApplicationContextAware {
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/AutoWiringSpringBeanQuartzTaskFactory.java]: no need for the interface

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTaskConfig.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	
+	@Value (CommonConstants.$SERVICE_REGISTRY_PING_SCHEDULED_WD)
+	private boolean pingScheduled;
+	
+	@Value (CommonConstants.$SERVICE_REGISTRY_PING_INTERVAL_WD)
+	private int pingInterval;
+	
+	private final int schedulerDelay = 10;
+	
+	private final String nameOfTrigger = "Providers_Reachability_Task_Trigger";
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTaskConfig.java]: static

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTaskConfig.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	private final String nameOfTask = "Providers_Reachability_Task_Detail";
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ProvidersReachabilityTaskConfig.java]: static

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTask.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	
+	protected Logger logger = LogManager.getLogger(ServiceEndOfValidityCheckTask.class);
+	private final int pageSize = 1000;
+	
+	@Autowired
+	private ServiceRegistryDBService serviceRegistryDBService;
+
+	@Override
+	public void execute(final JobExecutionContext context) throws JobExecutionException {
+		logger.info("STARTED: Services end of validity check task");
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTask.java]: debug

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTask.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+		final List<Long> removedServiceRegistryIDs = checkServicesEndOfValidity();
+		logger.info("FINISHED: Services end of validity check task. Number of removed service registry entry: " + removedServiceRegistryIDs.size());
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTask.java]: debug

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTask.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+		}
+		return removedServiceRegistryIDs;
+	}
+	
+	private List<Long> removeRegisteredServicesWithInvalidTTL(final Page<ServiceRegistry> pageOfServiceEntries) {
+		final List<Long> removedServiceRegistryIDs = new ArrayList<>();
+		for (final ServiceRegistry serviceRegistryEntry : pageOfServiceEntries) {
+			final ZonedDateTime endOfValidity = serviceRegistryEntry.getEndOfValidity();
+			if (endOfValidity != null && ! isTTLValid(endOfValidity)) {
+				serviceRegistryDBService.removeServiceRegistryEntryById(serviceRegistryEntry.getId());
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTask.java]: consider using bulk delete solution

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTaskConfig.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	
+	@Value (CommonConstants.$SERVICE_REGISTRY_TTL_SCHEDULED_WD)
+	private boolean ttlScheduled;
+	
+	@Value (CommonConstants.$SERVICE_REGISTRY_TTL_INTERVAL_WD)
+	private int ttlInterval;
+	
+	private final int schedulerDelay = 15;
+	
+	private final String nameOfTrigger = "Services_End_OF_Validity_Check_Task_Trigger";
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTaskConfig.java]: static

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTaskConfig.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	private final String nameOfTask = "Services_End_OF_Validity_Check_Task_Detail";
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTaskConfig.java]: static

File: serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTaskConfig.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+			logger.info("Services end of validity task is not adjusted");
+		}
+		return schedulerFactory;        
+    }
+	
+	@Bean
+    public SimpleTriggerFactoryBean servicesEndOfValidityCheckTaskTrigger() {
+		final SimpleTriggerFactoryBean trigger = new SimpleTriggerFactoryBean();
+		trigger.setJobDetail(servicesEndOfValidityCheckTaskDetail().getObject());
+        trigger.setRepeatInterval(ttlInterval * 60000);
```
Comment (rbocsi) [inline serviceregistry/src/main/java/eu/arrowhead/core/serviceregistry/quartz/task/ServiceEndOfValidityCheckTaskConfig.java]: maybe constant

File: serviceregistry/src/main/resources/application.properties [hunk 15]
```diff
@@ -25,18 +25,18 @@ server.port=8443
 # Service Registry has an optional feature to ping service providers in a fixed time interval,
 # and remove service offerings where the service provider was not available
 # use this feature (true/false)
-ping_scheduled=false
+ping_scheduled=true
 # how much time the Service Registry should wait for the ping response (in milliseconds)
 ping_timeout=5000
 # how frequently should the ping happen, in minutes
-ping_interval=60
+ping_interval=2
```
Comment (rbocsi) [inline serviceregistry/src/main/resources/application.properties]: try not to push testing settings to the repository

**Review by rbocsi — CHANGES_REQUESTED** [review]
see comments

**Comment by borditamas** [comment]
Changed as requested, please check again

**Review by rbocsi — APPROVED** [review]
seems ok

**Pull request merged**